File: illusionflow.c
#include <stdint.h>

struct illusionFlowState64 {
  uint64_t a;
  uint64_t b;
  uint64_t c;
  uint64_t d;
  uint64_t e;
  uint64_t f;
};

struct illusionFlowState32 {
  uint32_t a;
  uint32_t b;
  uint32_t c;
  uint32_t d;
  uint32_t e;
  uint32_t f;
};

struct illusionFlowState16 {
  uint16_t a;
  uint16_t b;
  uint16_t c;
  uint16_t d;
  uint16_t e;
  uint16_t f;
};

struct illusionFlowState8 {
  uint8_t a;
  uint8_t b;
  uint8_t c;
  uint8_t d;
  uint8_t e;
  uint8_t f;
};

uint64_t illusionFlow64(struct illusionFlowState64 *s) {
  uint64_t mix = (s->d << 25) | (s->d >> 39);

  s->d = s->c;
  s->c = s->b;
  s->f += s->e;
  s->e += s->b;
  s->b = s->a + 111111111111111111;
  s->a = s->f ^ mix;
  return s->a ^ (s->b >> 1);
}

uint32_t illusionFlow32(struct illusionFlowState32 *s) {
  uint32_t mix = (s->d << 13) | (s->d >> 19);

  s->d = s->c;
  s->c = s->b;
  s->f += s->e;
  s->e += s->b;
  s->b = s->a + 111111111;
  s->a = s->f ^ mix;
  return s->a ^ (s->b >> 1);
}

uint16_t illusionFlow16(struct illusionFlowState16 *s) {
  uint16_t mix = (s->d << 6) | (s->d >> 10);

  s->d = s->c;
  s->c = s->b;
  s->f += s->e;
  s->e += s->b;
  s->b = s->a + 1111;
  s->a = s->f ^ mix;
  return s->a ^ (s->b >> 1);
}

uint8_t illusionFlow8(struct illusionFlowState8 *s) {
  uint8_t mix = (s->d << 3) | (s->d >> 5);

  s->d = s->c;
  s->c = s->b;
  s->f += s->e;
  s->e += s->b;
  s->b = s->a + 111;
  s->a = s->f ^ mix;
  return s->a ^ (s->b >> 1);
}

File: illusionflow.cs
using System;

public class IllusionFlow64 {
  public ulong a;
  public ulong b;
  public ulong c;
  public ulong d;
  public ulong e;
  public ulong f;

  public ulong Next() {
    ulong mix = (d << 25) | (d >> 39);

    d = c;
    c = b;
    f += e;
    e += b;
    b = a + 111111111111111111;
    a = f ^ mix;
    return a ^ (b >> 1);
  }
}

public class IllusionFlow32 {
  public uint a;
  public uint b;
  public uint c;
  public uint d;
  public uint e;
  public uint f;

  public uint Next() {
    uint mix = (d << 13) | (d >> 19);

    d = c;
    c = b;
    f += e;
    e += b;
    b = a + 111111111;
    a = f ^ mix;
    return a ^ (b >> 1);
  }
}

public class IllusionFlow16 {
  public ushort a;
  public ushort b;
  public ushort c;
  public ushort d;
  public ushort e;
  public ushort f;

  public ushort Next() {
    ushort mix = (ushort) ((d << 6) | (d >> 10));

    d = c;
    c = b;
    f += e;
    e += b;
    b = (ushort) (a + 1111);
    a = (ushort) (f ^ mix);
    return (ushort) (a ^ (b >> 1));
  }
}

public class IllusionFlow8 {
  public byte a;
  public byte b;
  public byte c;
  public byte d;
  public byte e;
  public byte f;

  public byte Next() {
    byte mix = (byte) ((d << 3) | (d >> 5));

    d = c;
    c = b;
    f += e;
    e += b;
    b = (byte) (a + 111);
    a = (byte) (f ^ mix);
    return (byte) (a ^ (b >> 1));
  }
}
